### DISCUSSION TITLE
How to serve Template or do Redirect on Error? "Template / Redirect" object is not callable !

### DISCUSSION THREAD

**Discussion by 4n1qz5skwv** [litestar-org/litestar#1645] [Q&A] [answered]
**input**
```python
from pathlib import Path

from litestar import Litestar, Request, get
from litestar.contrib.jinja import JinjaTemplateEngine
from litestar.response_containers import Template, Redirect
from litestar.template.config import TemplateConfig


@get("/home")
async def aaa() -> str:
    return 'home!'


@get("/template")
async def bbb() -> Template:
    return Template(name="abc.html")


def show_template_on_error(_: Request, __: Exception) -> Template:
    return Template(name="abc.html")


@get("/template-error", exception_handlers={Exception: show_template_on_error})
async def ccc() -> str:
    1 / 0
    return ''


@get("/redirect")
async def ddd() -> Redirect:
    return Redirect(path='/home', status_code=302)


def redirect_on_error(_: Request, __: Exception) -> Redirect:
    return Redirect(path='/home', status_code=302)


@get("/redirect-error", exception_handlers={Exception: redirect_on_error})
async def eee() -> str:
    1 / 0
    return ''


app = Litestar(route_handlers=[aaa, bbb, ccc, ddd, eee],
               template_config=TemplateConfig(directory=Path(__file__).parent, engine=JinjaTemplateEngine))

if __name__ == '__main__':
    import uvicorn

    uvicorn.run('0:app')

```

**curl**
```cmd
Microsoft Windows [Version 10.0.19044.2846]
(c) Microsoft Corporation. All rights reserved.

C:\Users\User-PC>curl http://127.0.0.1:8000/template
<p> I am abc</p>
C:\Users\User-PC>curl http://127.0.0.1:8000/template-error
{"status_code":500,"detail":"TypeError(\"'Template' object is not callable\")"}
C:\Users\User-PC>curl http://127.0.0.1:8000/home
home!
C:\Users\User-PC>curl http://127.0.0.1:8000/redirect -L
home!
C:\Users\User-PC>curl http://127.0.0.1:8000/redirect-error
{"status_code":500,"detail":"TypeError(\"'Redirect' object is not callable\")"}
C:\Users\User-PC>
```

**logs**
```
INFO:     Started server process [1916]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:8000 (Press CTRL+C to quit)
INFO:     127.0.0.1:52573 - "GET /template HTTP/1.1" 200 OK
INFO:     127.0.0.1:52574 - "GET /template-error HTTP/1.1" 500 Internal Server Error
INFO:     127.0.0.1:52575 - "GET /home HTTP/1.1" 200 OK
INFO:     127.0.0.1:52576 - "GET /redirect HTTP/1.1" 302 Found
INFO:     127.0.0.1:52576 - "GET /home HTTP/1.1" 200 OK
INFO:     127.0.0.1:52577 - "GET /redirect-error HTTP/1.1" 500 Internal Server Error
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [1916]
```

**Comment by 4n1qz5skwv** [comment]
There's no example in the doc. Is this a Bug? or A limitation of the design?

**Comment by Goldziher** [comment, marked as answer]
An exception handler should return a response. Instead if `Template` return a `TemplateResponse`

**Reply by 4n1qz5skwv** [reply]
> An exception handler should return a response. Instead if `Template` return a `TemplateResponse`

```python
def show_template_on_error(_: Request, __: Exception) -> TemplateResponse:
    return TemplateResponse(template_name="abc.html",template_engine=JinjaTemplateEngine,status_code=200,context={})
```

```
{"status_code":500,"detail":"TypeError(\"JinjaTemplateEngine.get_template() missing 1 required positional argument: 'template_name'\")"}
```

Also, do we have pass these many values like context, engine (which already in app init as template_config) everywhere?

**Reply by 4n1qz5skwv** [reply]
**Edited:**

Might I humbly beseech your graciousness to kindly provide some examples, if you would be so kind and generous. As someone well-versed in the world of web frameworks, I am curious as to why certain standard practices, such as showing a 404 template or redirecting to the home page based on conditions, have not been clearly outlined in the documentation. Would it be possible for you to lend your expertise and guide me in this matter? Your assistance would be most appreciated, and I offer my sincere thanks in anticipation of your valuable insights.

**Reply by Goldziher** [reply]
```python
def show_template_on_error(request: Request, exc: Exception) -> Response:
    return Template(name="abc.html").to_response(
        headers={},
        media_type=MediaType.HTML,
        status_code=getattr(exc, "status_code", HTTP_400_BAD_REQUEST),
        app=request.app,
        request=request,
    )
```

**Reply by Goldziher** [reply]
As for a specific 404 helper, its a feature request, and you can request it. 

Anyhow, I do not particularly appreciate your tone. You opened a lot of issues and discussions, and you got a lot of attention. This is not the way to show appreciation of our time and efforts.

**Reply by provinzkraut** [reply]
@4n1qz5skwv 

> I don't know, why doc lags these standard important parts..

Because developing a web framework, especially one with a scope as large as Litestar's is a lot of work. Work none of use get paid for.

We are aware that documentation is still lacking in quite a few areas. Just like we are aware of a lot of other things that need attention. But we can only do so much with the time we have, and not everything can be of the highest priority. You might feel that, since you are hitting this particular issue, it should be a high priority thing, but imagine everyone else feeling the same way.

Since this is an open source project however, there is a very simple solution to fix something you think needs urgent fixing: You can contribute. We're always welcoming new contributors and are happy to guide you.

But it's important to remember that, while we are dedicated to this project and try our best to give every issue the attention it deserves, you are not our customer and this is not a paid product. We are giving away our labour for free here, to the benefit of the community. If you have an issue with the quality of this labour, you can contribute yourself and improve it.

**Reply by 4n1qz5skwv** [reply]
@Goldziher 

> You opened a lot of issues and discussions, and you got a lot of attention. This is not the way to show appreciation of our time and efforts.

 Okay I will quit since you don't seem to appreciate it when we mention issues. It's frustrating because we thought we were helping by reporting the issues we found.


-------


@provinzkraut 
> Because developing a web framework, especially one with a scope as large as Litestar's is a lot of work. Work none of use get paid for.

**Thank you for all you work and support. Truly appreciate for your time and efforts.**

**Reply by provinzkraut** [reply]
> Okay I will quit since you don't seem to appreciate it when we mention issues. It's frustrating because we thought we were helping by reporting the issues we found.

It's not *that* you report issue, it's rather how you go about it. We appreciate any contributions, including the reporting of issues. 

>  I don't know, why doc lags these standard important parts

This framing is what's problematic I'd say. Had you simply said "I did not see this mentioned in the docs, should I open an issue about this so it can be included?" it would have been perceived are more constructive and not just, well, naggy. 

I too am often finding myself in a situation where I think "oh, why isn't this mentioned in the docs?", and I can definitely understand that it can be frustrating. But if your comment about this seems to come from a place of understanding and good will, it will generally always be received better than if it's just a complaint.

<hr>

At least from my side, please feel free to continue raising issues, and finding bugs. As long as you do so with a friendly tone and cooperative approach, there's absolutely nothing wrong with it!

**Reply by Goldziher** [reply]
>  Okay I will quit since you don't seem to appreciate it when we mention issues. It's frustrating because we thought we were helping by reporting the issues we found.

You're not punishing me. If you want assistance and cooperation, be respectful. It's dead simple.

**Reply by 4n1qz5skwv** [reply]
>  be respectful

With all due respect and veneration, I humbly acknowledge that my previous message may have lacked the appropriate level of reverence and deference that your esteemed position deserves. I beg your indulgence and offer my sincerest apologies if any offense was taken. I understand that in our communication, it is of the utmost importance to maintain the highest standards of respect and deference, particularly when addressing an individual of your lofty standing. Rest assured that in the future, I will ensure that my messages exhibit the utmost respect and deference in all ways possible. Please accept my humblest apologies and deepest respects.
